proxy/server.go

```diff
@@ -18,74 +18,135 @@
 package proxy
 
 import (
-	"bufio"
-	"fmt"
+	"io"
+	"log"
 	"net"
-	"net/http"
+	"net/http/httputil"
 	"strings"
 
-	"github.com/elazarl/goproxy"
+	"github.com/inconshreveable/go-vhost"
+	"github.com/pkg/errors"
+	"github.com/soheilhy/cmux"
 	netproxy "golang.org/x/net/proxy"
 )
 
-type stickyHasher interface {
-	Hash(ip string) string
-}
-
 type domainTracker interface {
 	Inc(domain string)
 }
 
+type proxyServer struct {
+	addr   string
+	dialer netproxy.Dialer
+	sm     *stickyMapper
+	dt     domainTracker
+}
+
 // NewServer returns new instance of HTTP transparent proxy server
-func NewServer(addr string, upstreamDialer netproxy.Dialer, mapper stickyHasher, dt domainTracker) *http.Server {
-	server := goproxy.NewProxyHttpServer()
-	server.Verbose = true
-	server.NonproxyHandler = http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		if req.Host == "" {
-			fmt.Fprintln(w, "Cannot handle requests without Host header, e.g., HTTP 1.0")
-			return
-		}
-		req.URL.Scheme = "http"
-		req.URL.Host = req.Host
-		server.ServeHTTP(w, req)
-	})
-
-	server.ConnectDial = upstreamDialer.Dial
-	server.OnRequest().HandleConnectFunc(func(host string, ctx *goproxy.ProxyCtx) (*goproxy.ConnectAction, string) {
-		return goproxy.OkConnect, host
-	})
-
-	server.OnRequest().DoFunc(func(req *http.Request, ctx *goproxy.ProxyCtx) (*http.Request, *http.Response) {
-		conn, err := upstreamDialer.Dial("tcp", req.Host+":80")
-		if err != nil {
-			return req, nil
-		}
+func NewServer(addr string, upstreamDialer netproxy.Dialer, mapper *stickyMapper, dt domainTracker) *proxyServer {
+	return &proxyServer{
+		addr:   addr,
+		dialer: upstreamDialer,
+		sm:     mapper,
+		dt:     dt,
+	}
+}
+
+func (s *proxyServer) ListenAndServe() {
+	ln, err := net.Listen("tcp", s.addr)
+	if err != nil {
+		log.Fatalf("Error listening for https connections - %v", err)
+	}
 
-		domain := strings.Split(req.Host, ":")
-		dt.Inc(domain[0])
+	m := cmux.New(ln)
 
-		if proxyConnection, ok := conn.(*Connection); ok {
-			clientHost, _, _ := net.SplitHostPort(req.RemoteAddr)
-			if err := proxyConnection.ConnectTo(conn, req.Host+":80", mapper.Hash(clientHost)); err != nil {
-				return req, nil
-			}
-		}
+	httpsL := m.Match(cmux.TLS())
+	httpL := m.Match(cmux.HTTP1Fast())
 
-		err = req.Write(conn)
-		if err != nil {
-			return req, nil
-		}
+	go s.handler(httpL, s.serveHTTP)
+	go s.handler(httpsL, s.serveTLS)
+
+	m.Serve()
+}
 
-		resp, err := http.ReadResponse(bufio.NewReader(conn), req)
+func (s *proxyServer) handler(l net.Listener, f func(c net.Conn)) {
+	for {
+		c, err := l.Accept()
```

Comment:
Some resource clean e.g. `defer c.Close()` needed.
Would be nice next to place where resources are allocated.

Reply:
Added closing it close to the `Accept`.